markdown: Add support for user group silent mention and other followups.

akshatdalton opened the pull request:
This will add support for user group silent mention and provides typeahead for them.

I see no apparent reason to support the typeahead for wildcard silent mentions. Therefore I've not added it.

Fixes: #11711.

timabbott commented on static/js/markdown.js:
```diff
@@ -102,8 +102,16 @@ export function apply_markdown(message) {
+
+                return `<span class="user-mention silent" data-user-id="*">${_.escape(
+                    mention,
+                )}</span>`;
```
I think we want to rather than having the whole syntax be an if/else just have a variable `classes = "user-mention"` that has ` silent` added to it for the silent case.  That way, we avoid duplicating the escaping logic (etc.), which would increase the risk of future bugs.

akshatdalton replied:
Done.

timabbott commented:
I merged the first half of this PR as the series ending with 9e5f2f841f28e936f788b2ddb03cef9c1925dcad; posted one comment that applies to 3 fo the remaining commits.

Otherwise this lgtm, thanks for the cleanly written commit series @akshatdalton!

akshatdalton commented:
Updated the PR; used `classes = "user-mention"` for wildcard mention and `classes = "user-mention-group"` for user-group mention. Also added another commit: https://github.com/zulip/zulip/pull/18529/commits/daed13b24504dff225a23ee867b59239931f648b which fixes `get_mention_syntax` function when `full_name `matches wildcard.

@timabbott, please review.

timabbott commented on static/js/markdown.js:
```diff
@@ -187,15 +187,18 @@ export function apply_markdown(message) {
+                const classes = silently ? "user-group-mention silent" : "user-group-mention";
+                if (helpers.is_member_of_user_group(group.id, helpers.my_user_id()) && !silently) {
                     message.mentioned = true;
                 }
```
This is hard to read.  I think it'd be a lot better to just have a single `if (silently)` and set `group_name` or classes within that, than having 3 checks for silently in inline conditionals.

akshatdalton replied:
I initially did this, but I don't know why I thought I shouldn't expand the code :facepalm: and try to keep it compact.
Thanks! for explaining.

timabbott commented on static/js/markdown.js:
```diff
@@ -210,6 +210,15 @@ export function apply_markdown(message) {
             });
+
+            // Silence quoted user group mentions.
+            const user_group_re = /<span.*user-group-mention.*data-user-group-id="(\d+)"[^>]*>@/gm;
```
This regex seems likely wrong, in that you have a `.*` that could match a bunch of characters including another `><`, i.e. end up matching across multiple adjacent spans in a weird way

It appears the previous regex has the same bug; I think a `[^>]*` would work for the two `.*` blocks in the middle, so maybe do a commit fixing the existing instance and then fix this?

akshatdalton replied:
Fixed the regex for this block and the previous block.

akshatdalton commented:
- Fixed the regex as mentioned in https://github.com/zulip/zulip/pull/18529#discussion_r635504056.
- Simplified the code as mentioned in https://github.com/zulip/zulip/pull/18529#discussion_r635502251.

@timabbott, please review.

timabbott commented on static/js/markdown.js:
```diff
@@ -101,8 +101,18 @@ export function apply_markdown(message) {
+                    message.mentioned = true;
+                    display_text = "@" + mention;
+                    classes = "user-mention";
+                }
```
I found it a lot more readable with this model, where we set all the parameters in each conditional, so I made that tweak and merged this.

Thanks @akshatdalton!

timabbott commented:
Merged, thanks @akshatdalton!  I did fix a couple typos in the commit messages; please take some extra time to proofread writing in commit messages and comments, since finding and fixing typos adds significant time to reviewing otherwise easily merged PRs :)

It's nice to have a half-dozen little issues fixed together here!

akshatdalton commented:
Sorry for the inconvenience; I'll take extra care the next time.
